File: CSharp/src/Runtime/Scripts/XLog/tag.cs
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

using System;
using System.Collections.Generic;
using System.Text;

namespace FrameworkEase.Universal.X.Utility
{
    public partial class XLog
    {
        public struct Tag
        {
            internal List<string> pairs;

            public readonly int Count => pairs != null ? pairs.Count / 2 : 0;

            public void Set(string key, string value)
            {
                pairs ??= new List<string>();
                for (var i = 0; i < pairs.Count; i += 2)
                {
                    if (pairs[i] != key) continue;
                    pairs[i + 1] = value;
                    return;
                }
                pairs.Add(key);
                pairs.Add(value);
            }

            public readonly string Get(string key)
            {
                if (pairs == null) return string.Empty;
                for (var i = 0; i < pairs.Count; i += 2)
                {
                    if (pairs[i] == key) return pairs[i + 1];
                }
                return string.Empty;
            }

            public readonly void Range(Func<string, string, bool> callback)
            {
                if (callback == null || pairs == null) return;
                for (var i = 0; i < pairs.Count; i += 2)
                {
                    if (!callback(pairs[i], pairs[i + 1])) break;
                }
            }

            public readonly string Stringify()
            {
                if (pairs == null || pairs.Count == 0) return string.Empty;
                var builder = new StringBuilder();
                builder.Append("[");
                var first = true;
                for (var i = 0; i < pairs.Count; i += 2)
                {
                    if (!first) builder.Append(", ");
                    else first = false;
                    builder.Append(pairs[i]);
                    builder.Append("=");
                    builder.Append(pairs[i + 1]);
                }
                builder.Append("]");
                return builder.ToString();
            }
        }
    }
}

File: Go/XLog/tag.go
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

package XLog

import (
	"fmt"
	"strings"
)

type Tag struct {
	pairs []string
}

func (tag *Tag) Count() int { return len(tag.pairs) / 2 }

func (tag *Tag) Set(key string, value string) {
	for i := 0; i < len(tag.pairs); i += 2 {
		if tag.pairs[i] == key {
			tag.pairs[i+1] = value
			return
		}
	}
	tag.pairs = append(tag.pairs, key, value)
}

func (tag *Tag) Get(key string) string {
	for i := 0; i < len(tag.pairs); i += 2 {
		if tag.pairs[i] == key {
			return tag.pairs[i+1]
		}
	}
	return ""
}

func (tag *Tag) Range(callback func(key string, value string) bool) {
	if callback == nil || len(tag.pairs) == 0 {
		return
	}
	for i := 0; i < len(tag.pairs); i += 2 {
		if !callback(tag.pairs[i], tag.pairs[i+1]) {
			break
		}
	}
}

func (tag *Tag) Stringify() string {
	if len(tag.pairs) == 0 {
		return ""
	}
	var builder strings.Builder
	builder.WriteString("[")
	first := true
	for i := 0; i < len(tag.pairs); i += 2 {
		if !first {
			builder.WriteString(", ")
		} else {
			first = false
		}
		fmt.Fprintf(&builder, "%s=%s", tag.pairs[i], tag.pairs[i+1])
	}
	builder.WriteString("]")
	return builder.String()
}
